File: clients/kotlin/src/main/kotlin/model/Properties.kt
package model

import util.StreamUtil

class Properties {
    var maxTickCount: Int = 0
    var teamSize: Int = 0
    var ticksPerSecond: Double = 0.0
    var updatesPerTick: Int = 0
    lateinit var lootBoxSize: model.Vec2Double
    lateinit var unitSize: model.Vec2Double
    var unitMaxHorizontalSpeed: Double = 0.0
    var unitFallSpeed: Double = 0.0
    var unitJumpTime: Double = 0.0
    var unitJumpSpeed: Double = 0.0
    var jumpPadJumpTime: Double = 0.0
    var jumpPadJumpSpeed: Double = 0.0
    var unitMaxHealth: Int = 0
    var healthPackHealth: Int = 0
    lateinit var weaponParams: MutableMap<model.WeaponType, model.WeaponParams>
    lateinit var mineSize: model.Vec2Double
    lateinit var mineExplosionParams: model.ExplosionParams
    var minePrepareTime: Double = 0.0
    var mineTriggerTime: Double = 0.0
    var mineTriggerRadius: Double = 0.0
    var killScore: Int = 0
    constructor() {}
    constructor(maxTickCount: Int, teamSize: Int, ticksPerSecond: Double, updatesPerTick: Int, lootBoxSize: model.Vec2Double, unitSize: model.Vec2Double, unitMaxHorizontalSpeed: Double, unitFallSpeed: Double, unitJumpTime: Double, unitJumpSpeed: Double, jumpPadJumpTime: Double, jumpPadJumpSpeed: Double, unitMaxHealth: Int, healthPackHealth: Int, weaponParams: MutableMap<model.WeaponType, model.WeaponParams>, mineSize: model.Vec2Double, mineExplosionParams: model.ExplosionParams, minePrepareTime: Double, mineTriggerTime: Double, mineTriggerRadius: Double, killScore: Int) {
        this.maxTickCount = maxTickCount
        this.teamSize = teamSize
        this.ticksPerSecond = ticksPerSecond
        this.updatesPerTick = updatesPerTick
        this.lootBoxSize = lootBoxSize
        this.unitSize = unitSize
        this.unitMaxHorizontalSpeed = unitMaxHorizontalSpeed
        this.unitFallSpeed = unitFallSpeed
        this.unitJumpTime = unitJumpTime
        this.unitJumpSpeed = unitJumpSpeed
        this.jumpPadJumpTime = jumpPadJumpTime
        this.jumpPadJumpSpeed = jumpPadJumpSpeed
        this.unitMaxHealth = unitMaxHealth
        this.healthPackHealth = healthPackHealth
        this.weaponParams = weaponParams
        this.mineSize = mineSize
        this.mineExplosionParams = mineExplosionParams
        this.minePrepareTime = minePrepareTime
        this.mineTriggerTime = mineTriggerTime
        this.mineTriggerRadius = mineTriggerRadius
        this.killScore = killScore
    }
    companion object {
        @Throws(java.io.IOException::class)
        fun readFrom(stream: java.io.InputStream): Properties {
            val result = Properties()
            result.maxTickCount = StreamUtil.readInt(stream)
            result.teamSize = StreamUtil.readInt(stream)
            result.ticksPerSecond = StreamUtil.readDouble(stream)
            result.updatesPerTick = StreamUtil.readInt(stream)
            result.lootBoxSize = model.Vec2Double.readFrom(stream)
            result.unitSize = model.Vec2Double.readFrom(stream)
            result.unitMaxHorizontalSpeed = StreamUtil.readDouble(stream)
            result.unitFallSpeed = StreamUtil.readDouble(stream)
            result.unitJumpTime = StreamUtil.readDouble(stream)
            result.unitJumpSpeed = StreamUtil.readDouble(stream)
            result.jumpPadJumpTime = StreamUtil.readDouble(stream)
            result.jumpPadJumpSpeed = StreamUtil.readDouble(stream)
            result.unitMaxHealth = StreamUtil.readInt(stream)
            result.healthPackHealth = StreamUtil.readInt(stream)
            val weaponParamsSize = StreamUtil.readInt(stream)
            result.weaponParams = mutableMapOf()
            for (i in 0 until weaponParamsSize) {
                var weaponParamsKey: model.WeaponType
                when (StreamUtil.readInt(stream)) {
                0 ->weaponParamsKey = model.WeaponType.PISTOL
                1 ->weaponParamsKey = model.WeaponType.ASSAULT_RIFLE
                2 ->weaponParamsKey = model.WeaponType.ROCKET_LAUNCHER
                else -> throw java.io.IOException("Unexpected discriminant value")
                }
                var weaponParamsValue: model.WeaponParams
                weaponParamsValue = model.WeaponParams.readFrom(stream)
                result.weaponParams.put(weaponParamsKey, weaponParamsValue)
            }
            result.mineSize = model.Vec2Double.readFrom(stream)
            result.mineExplosionParams = model.ExplosionParams.readFrom(stream)
            result.minePrepareTime = StreamUtil.readDouble(stream)
            result.mineTriggerTime = StreamUtil.readDouble(stream)
            result.mineTriggerRadius = StreamUtil.readDouble(stream)
            result.killScore = StreamUtil.readInt(stream)
            return result
        }
    }
    @Throws(java.io.IOException::class)
    fun writeTo(stream: java.io.OutputStream) {
        StreamUtil.writeInt(stream, maxTickCount)
        StreamUtil.writeInt(stream, teamSize)
        StreamUtil.writeDouble(stream, ticksPerSecond)
        StreamUtil.writeInt(stream, updatesPerTick)
        lootBoxSize.writeTo(stream)
        unitSize.writeTo(stream)
        StreamUtil.writeDouble(stream, unitMaxHorizontalSpeed)
        StreamUtil.writeDouble(stream, unitFallSpeed)
        StreamUtil.writeDouble(stream, unitJumpTime)
        StreamUtil.writeDouble(stream, unitJumpSpeed)
        StreamUtil.writeDouble(stream, jumpPadJumpTime)
        StreamUtil.writeDouble(stream, jumpPadJumpSpeed)
        StreamUtil.writeInt(stream, unitMaxHealth)
        StreamUtil.writeInt(stream, healthPackHealth)
        StreamUtil.writeInt(stream, weaponParams.size)
        for (weaponParamsEntry in weaponParams) {
            StreamUtil.writeInt(stream, weaponParamsEntry.key.discriminant)
            weaponParamsEntry.value.writeTo(stream)
        }
        mineSize.writeTo(stream)
        mineExplosionParams.writeTo(stream)
        StreamUtil.writeDouble(stream, minePrepareTime)
        StreamUtil.writeDouble(stream, mineTriggerTime)
        StreamUtil.writeDouble(stream, mineTriggerRadius)
        StreamUtil.writeInt(stream, killScore)
    }
}

File: clients/javascript/model/properties.js
const Vec2Double = require('./vec2-double').Vec2Double;
const WeaponParams = require('./weapon-params').WeaponParams;
const ExplosionParams = require('./explosion-params').ExplosionParams;
class Properties {
    constructor(maxTickCount, teamSize, ticksPerSecond, updatesPerTick, lootBoxSize, unitSize, unitMaxHorizontalSpeed, unitFallSpeed, unitJumpTime, unitJumpSpeed, jumpPadJumpTime, jumpPadJumpSpeed, unitMaxHealth, healthPackHealth, weaponParams, mineSize, mineExplosionParams, minePrepareTime, mineTriggerTime, mineTriggerRadius, killScore) {
        this.maxTickCount = maxTickCount;
        this.teamSize = teamSize;
        this.ticksPerSecond = ticksPerSecond;
        this.updatesPerTick = updatesPerTick;
        this.lootBoxSize = lootBoxSize;
        this.unitSize = unitSize;
        this.unitMaxHorizontalSpeed = unitMaxHorizontalSpeed;
        this.unitFallSpeed = unitFallSpeed;
        this.unitJumpTime = unitJumpTime;
        this.unitJumpSpeed = unitJumpSpeed;
        this.jumpPadJumpTime = jumpPadJumpTime;
        this.jumpPadJumpSpeed = jumpPadJumpSpeed;
        this.unitMaxHealth = unitMaxHealth;
        this.healthPackHealth = healthPackHealth;
        this.weaponParams = weaponParams;
        this.mineSize = mineSize;
        this.mineExplosionParams = mineExplosionParams;
        this.minePrepareTime = minePrepareTime;
        this.mineTriggerTime = mineTriggerTime;
        this.mineTriggerRadius = mineTriggerRadius;
        this.killScore = killScore;
    }
    static async readFrom(stream) {
        let maxTickCount;
        maxTickCount = await stream.readInt();
        let teamSize;
        teamSize = await stream.readInt();
        let ticksPerSecond;
        ticksPerSecond = await stream.readDouble();
        let updatesPerTick;
        updatesPerTick = await stream.readInt();
        let lootBoxSize;
        lootBoxSize = await Vec2Double.readFrom(stream);
        let unitSize;
        unitSize = await Vec2Double.readFrom(stream);
        let unitMaxHorizontalSpeed;
        unitMaxHorizontalSpeed = await stream.readDouble();
        let unitFallSpeed;
        unitFallSpeed = await stream.readDouble();
        let unitJumpTime;
        unitJumpTime = await stream.readDouble();
        let unitJumpSpeed;
        unitJumpSpeed = await stream.readDouble();
        let jumpPadJumpTime;
        jumpPadJumpTime = await stream.readDouble();
        let jumpPadJumpSpeed;
        jumpPadJumpSpeed = await stream.readDouble();
        let unitMaxHealth;
        unitMaxHealth = await stream.readInt();
        let healthPackHealth;
        healthPackHealth = await stream.readInt();
        let weaponParams;
        weaponParams = new Map();
        for (let i = await stream.readInt(); i > 0; i--) {
            let weaponParamsKey;
            let weaponParamsValue;
            weaponParamsKey = stream.readInt();
            weaponParamsValue = await WeaponParams.readFrom(stream);
            weaponParams.set(weaponParamsKey, weaponParamsValue);
        }
        let mineSize;
        mineSize = await Vec2Double.readFrom(stream);
        let mineExplosionParams;
        mineExplosionParams = await ExplosionParams.readFrom(stream);
        let minePrepareTime;
        minePrepareTime = await stream.readDouble();
        let mineTriggerTime;
        mineTriggerTime = await stream.readDouble();
        let mineTriggerRadius;
        mineTriggerRadius = await stream.readDouble();
        let killScore;
        killScore = await stream.readInt();
        return new Properties(maxTickCount, teamSize, ticksPerSecond, updatesPerTick, lootBoxSize, unitSize, unitMaxHorizontalSpeed, unitFallSpeed, unitJumpTime, unitJumpSpeed, jumpPadJumpTime, jumpPadJumpSpeed, unitMaxHealth, healthPackHealth, weaponParams, mineSize, mineExplosionParams, minePrepareTime, mineTriggerTime, mineTriggerRadius, killScore);
    }
    async writeTo(stream) {
        let maxTickCount = this.maxTickCount;
        await stream.writeInt(maxTickCount);
        let teamSize = this.teamSize;
        await stream.writeInt(teamSize);
        let ticksPerSecond = this.ticksPerSecond;
        await stream.writeDouble(ticksPerSecond);
        let updatesPerTick = this.updatesPerTick;
        await stream.writeInt(updatesPerTick);
        let lootBoxSize = this.lootBoxSize;
        await lootBoxSize.writeTo(stream);
        let unitSize = this.unitSize;
        await unitSize.writeTo(stream);
        let unitMaxHorizontalSpeed = this.unitMaxHorizontalSpeed;
        await stream.writeDouble(unitMaxHorizontalSpeed);
        let unitFallSpeed = this.unitFallSpeed;
        await stream.writeDouble(unitFallSpeed);
        let unitJumpTime = this.unitJumpTime;
        await stream.writeDouble(unitJumpTime);
        let unitJumpSpeed = this.unitJumpSpeed;
        await stream.writeDouble(unitJumpSpeed);
        let jumpPadJumpTime = this.jumpPadJumpTime;
        await stream.writeDouble(jumpPadJumpTime);
        let jumpPadJumpSpeed = this.jumpPadJumpSpeed;
        await stream.writeDouble(jumpPadJumpSpeed);
        let unitMaxHealth = this.unitMaxHealth;
        await stream.writeInt(unitMaxHealth);
        let healthPackHealth = this.healthPackHealth;
        await stream.writeInt(healthPackHealth);
        let weaponParams = this.weaponParams;
        await stream.writeInt(weaponParams.size);
        for (let weaponParamsEntry of weaponParams) {
            let weaponParamsKey = weaponParamsEntry[0];
            let weaponParamsValue = weaponParamsEntry[1];
            await stream.writeInt(weaponParamsKey);
            await weaponParamsValue.writeTo(stream);
        }
        let mineSize = this.mineSize;
        await mineSize.writeTo(stream);
        let mineExplosionParams = this.mineExplosionParams;
        await mineExplosionParams.writeTo(stream);
        let minePrepareTime = this.minePrepareTime;
        await stream.writeDouble(minePrepareTime);
        let mineTriggerTime = this.mineTriggerTime;
        await stream.writeDouble(mineTriggerTime);
        let mineTriggerRadius = this.mineTriggerRadius;
        await stream.writeDouble(mineTriggerRadius);
        let killScore = this.killScore;
        await stream.writeInt(killScore);
    }
}
module.exports = { Properties: Properties }
